File: app/src/main/java/io/github/ovso/psytest/ui/main/fragment/adapter/VideoAdapter.java
package io.github.ovso.psytest.ui.main.fragment.adapter;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.ViewGroup;
import io.github.ovso.psytest.data.network.model.SearchItem;
import io.github.ovso.psytest.ui.base.interfaces.OnRecyclerViewItemClickListener;
import java.util.ArrayList;
import java.util.List;
import lombok.Setter;
import org.jetbrains.annotations.Nullable;

public class VideoAdapter extends RecyclerView.Adapter<VideoViewHolder>
    implements VideoAdapterView, VideoAdapterDataModel<SearchItem> {
  @Setter protected OnRecyclerViewItemClickListener<SearchItem> onRecyclerViewItemClickListener;
  private List<SearchItem> items = new ArrayList<>();

  @NonNull @Override
  public VideoViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
    return VideoViewHolder.Companion.create(viewGroup);
  }

  @Override
  public void onBindViewHolder(@NonNull VideoViewHolder viewHolder, int position) {
    viewHolder.bind(getItem(position));
    viewHolder.setOnItemClickListener(onRecyclerViewItemClickListener);
  }

  @Override public int getItemCount() {
    return getSize();
  }

  @Override public int getSize() {
    return items.size();
  }

  @Override public SearchItem getItem(int position) {
    return items.get(position);
  }

  @Override public void clear() {
    items.clear();
  }

  @Override public void refresh() {
    notifyDataSetChanged();
  }

  @Override public void refresh(int positionStart, int itemCount) {
    notifyItemRangeInserted(positionStart, itemCount);
  }

  @Override public void addAll(@Nullable List<? extends SearchItem> $items) {
    if ($items != null) {
      items.addAll($items);
    }
  }
}

File: app/src/main/java/io/github/ovso/psytest/ui/main/fragment/adapter/VideoAdapter.kt
package io.github.ovso.psytest.ui.main.fragment.adapter

import android.support.v7.widget.RecyclerView
import android.view.ViewGroup
import io.github.ovso.psytest.data.network.model.SearchItem
import io.github.ovso.psytest.ui.base.interfaces.OnRecyclerViewItemClickListener
import java.util.ArrayList

class VideoAdapter : RecyclerView.Adapter<VideoViewHolder>(),
    VideoAdapterView,
    VideoAdapterDataModel<SearchItem> {
  var onRecyclerViewItemClickListener: OnRecyclerViewItemClickListener<SearchItem>? = null
  private val items = ArrayList<SearchItem>()

  override val size: Int
    get() = items.size

  override fun onCreateViewHolder(
    viewGroup: ViewGroup,
    viewType: Int
  ): VideoViewHolder {
    return VideoViewHolder.create(viewGroup)
  }

  override fun onBindViewHolder(
    viewHolder: VideoViewHolder,
    position: Int
  ) {
    viewHolder.bind(getItem(position))
    viewHolder.onItemClickListener = onRecyclerViewItemClickListener
  }

  override fun getItemCount(): Int {
    return size
  }

  override fun getItem(position: Int): SearchItem {
    return items[position]
  }

  override fun clear() {
    items.clear()
  }

  override fun refresh() {
    notifyDataSetChanged()
  }

  override fun refresh(
    positionStart: Int,
    itemCount: Int
  ) {
    notifyItemRangeInserted(positionStart, itemCount)
  }

  override fun addAll(items: List<SearchItem>?) {
    this.items.addAll(items!!)
  }
}